packages/go_router/lib/src/route.dart

```diff
@@ -279,3 +317,62 @@ class GoRoute {
   ) =>
       const SizedBox.shrink();
 }
+
+/// A route that displays a UI shell around the matching child route. Builds
+/// a new Navigator that is used to display any matching sub-routes, instead
+/// of placing them on the root Navigator.
+///
+/// To display a child route on a different Navigator, provide it with a
+/// [parentNavigatorKey] that matches the key provided to either the [GoRouter] or
+/// [ShellRoute] constructor.
+///
+/// The widget built by the matching sub-route becomes the child parameter
+/// of the [builder].
+///
+/// For example:
+///
+/// ```
+/// ShellRoute(
+///   path: '/',
+///   builder: (BuildContext context, GoRouterState state, Widget child) {
+///     return Scaffold(
+///       appBar: AppBar(
+///         title: Text('App Shell')
+///       ),
+///       body: Center(
+///         child: child,
+///       ),
+///     );
+///   },
+///   routes: [
+///     GoRoute(
+///       path: 'a'
+///       builder: (BuildContext context, GoRouterState state) {
+///         return Text('Child Route "/a"');
+///       }
+///     ),
+///   ],
+/// ),
+/// ```
+///
+class ShellRoute extends RouteBase {
+  /// Constructs a [ShellRoute].
+  ShellRoute({
+    this.builder,
+    this.pageBuilder,
+    super.routes,
```

Comment:
we should assert the routes must not be empty.

Reply:
9be299a4425fd18707ebf35ee9eae024ef0d7734